feat(social-media-share): serve different static index.html files based on path

LinHuiqing opened the pull request:
## Problem

Social media crawlers are only able to see initial files served statically by server (i.e. `index.html`). To make our website more shareable on social media, these initial files should be populated with titles, meta description and open graph tags.

Related to #394 

## Solution

Get `server/src/bootstrap/index.ts` to modify `index.html` before serving them.

**Features**:

- (server/src/bootstrap/index.ts) send modified index.html for /agency/:shortname and /questions/:id paths for production

**Improvements**:

- (client/public/index.html) added og tags

## Before & After Screenshots

**BEFORE**:

Social media link preview:

<img width="529" alt="Screenshot 2021-10-05 at 5 07 44 PM" src="https://user-images.githubusercontent.com/37061143/135994304-7eadefdd-77b4-417d-89f8-cced2eff953e.png">

**AFTER**:

Main AskGov page:

<img width="528" alt="Screenshot 2021-10-05 at 5 13 29 PM" src="https://user-images.githubusercontent.com/37061143/135995145-d75ab4e6-98fc-4649-a836-bbfe5fe66fde.png">

Agency-specific page:

<img width="528" alt="Screenshot 2021-10-05 at 5 13 56 PM" src="https://user-images.githubusercontent.com/37061143/135995234-9fe125b5-4758-4723-aa42-85465fae9e7d.png">

Post page:

<img width="991" alt="Screenshot 2021-10-05 at 5 14 58 PM" src="https://user-images.githubusercontent.com/37061143/135995408-1ea3fa3e-a42d-446e-8fd1-c8fddd28cb03.png">

## Tests

Use [Facebook Sharing Debugger](https://developers.facebook.com/tools/debug/) to check that the title, meta description and og tags are populated. Specifically, test this for the following pages:

1. Main AskGov page
2. Agency-specific page
3. Post page

## Deploy Notes

**New dependencies**:

- `dependency` on `server`: 
  - `sanitize-html`
  - `cheerio`

**New dev dependencies**:

- `dependency` on `server`:  `@types/sanitize-html`

LinHuiqing committed:
feat(cheerio): serve different static index.html files based on path

- (server/package.json, server/package-lock.json) install cheerio
- (client/public/index.html) added og tags
- (server/src/bootstrap/index.ts) send modified index.html for /agency/:shortname and /questions/:id paths for production

LinHuiqing committed:
fix(desc-index.html): sanitize answer html for descriptions

- (server/package.json, server/package-lock.json) install sanitize-html and @types/sanitize-html
-
sanitize html of answer before populating description

LinHuiqing committed:
fix(image-index.html): remove populating of agency logo due to size

add og:image only when we have bigger agency logos ready

LinHuiqing committed:
fix(og-url): populate correct og:url content

LinHuiqing commented:
To be honest, `og:url` is still missing for some reason :/ not quite sure why or how to debug that?

LoneRifle commented on server/src/bootstrap/index.ts:
```diff
@@ -194,9 +198,100 @@ if (baseConfig.nodeEnv === Environment.Prod) {
+        shortname: req.params.shortname,
+      })
+    }
+    ;(async () => {
+      const agency = await agencyPromise()
+      if (agency.isOk()) {
+        $('meta[name="description"]').attr(
+          'content',
+          `Answers from ${
+            agency.value.longname
+          } (${req.params.shortname.toUpperCase()})`,
+        )
+        $('meta[property="og:description"]').attr(
+          'content',
+          `Answers from ${
+            agency.value.longname
+          } (${req.params.shortname.toUpperCase()})`,
+        )
+      } else {
+        $('meta[name="description"]').attr(
+          'content',
+          `Answers from ${req.params.shortname.toUpperCase()}`,
+        )
+        $('meta[property="og:description"]').attr(
+          'content',
+          `Answers from ${req.params.shortname.toUpperCase()}`,
+        )
+      }
+      res.header('content-type', 'text/html').send($.html())
+    })()
```
Why wrap it up like this? You can simply make the `(req, res) => { ... }` handler async

LoneRifle commented on server/src/bootstrap/index.ts:
```diff
@@ -194,9 +198,100 @@ if (baseConfig.nodeEnv === Environment.Prod) {
+    })()
+  })
+
+  app.get('*', (req, res) => {
+    const $ = cheerio.load(index)
+    $('meta[property="og:url"]').attr(
+      'content',
+      `${req.protocol}://${req.hostname}${req.originalUrl.toLowerCase()}`,
+    )
+    res.header('content-type', 'text/html').send($.html())
+  })
```
I think this is slightly spurious, so maybe we can drop this =)

LinHuiqing replied:
okays

LoneRifle reviewed (changes requested):
Final thoughts:
- At this point, the implementation is fairly involved. Per earlier comments I made the other time, extract this out into its own module, following the structure as you find in `modules/`. So perhaps you could define something like `modules/web` and go from there with controller, route, service etc
- What do you do when the question or agency does not exist?

LinHuiqing commented:
> * What do you do when the question or agency does not exist?

For this, what should we see actually? Should it be the default or some kind of "not found" page? Currently not found pages give the default preview too :/ 

Tangent: https://staging.ask.gov.sg/agency/no gives a blank page rather than a not found page. This is a bug, right?

LoneRifle commented:
It's a missing feature, to be more precise. I would think that you could basically serve the not found page when you can't find the agency or question, and return http status code 404, thereby also solving #256

LinHuiqing committed:
refactor(social-media-share): refactor serving of static files

- extract static page generation logic into web module

LoneRifle commented on server/src/routes/index.ts:
```diff
@@ -58,5 +61,6 @@ export const api = (options: ApiRouterOptions): express.Router => {
   router.use('/posts/answers', routeAnswers(options.answers))
   router.use('/agencies', routeAgencies({ controller: options.agency }))
   router.use('/enquiries', routeEnquiries({ controller: options.enquiries }))
+  router.use('/web', routeWeb({ controller: options.web }))
```
This isn't an api route..

LinHuiqing committed:
feat(404): redirect all invalid paths to not-found page

- redirect all invalid paths to /not-found
- redirect /questions/:id where question with id does not exist to /not-found
- redirect /agency/:shortname where agency with shortname does not exist to /not-found

LoneRifle committed:
fix(post): skip view count update if from web controller

LoneRifle committed:
fix(web): have proper logging of errors
